mlflow/models/model.py

```diff
@@ -747,9 +763,47 @@ def log(
                     await_registration_for=await_registration_for,
                     local_model_path=local_path,
                 )
-        model_info = mlflow_model.get_model_info()
-        if registered_model is not None:
-            model_info.registered_model_version = registered_model.version
+            model_info = mlflow_model.get_model_info()
+            if registered_model is not None:
+                model_info.registered_model_version = registered_model.version
+
+            # validate input example works for serving when logging the model
+            serving_input = mlflow_model.get_serving_input(local_path)
+            if mlflow_model.signature is None and serving_input is None:
+                _logger.warning(
+                    "Input example should be provided to infer model signature if the model "
+                    "signature is not provided when logging the model."
+                )
+            if serving_input:
+                try:
+                    # sklearn model might not have python_function flavor if it
+                    # doesn't define a predict function.
+                    # we should not fail if pyfunc model does not exist as the
+                    # model can not be served anyways
+                    pyfunc_model = mlflow.pyfunc.load_model(model_info.model_uri)
+                except Exception as e:
+                    _logger.debug(
+                        f"Failed to load model for serving validation: {e}", exc_info=True
+                    )
+                else:
+                    from mlflow.pyfunc.scoring_server import _parse_json_data
+
+                    try:
+                        parsed_input = _parse_json_data(
+                            serving_input,
+                            pyfunc_model.metadata,
+                            pyfunc_model.metadata.get_input_schema(),
+                        )
+                        pyfunc_model.predict(parsed_input.data, params=parsed_input.params)
+                    except Exception as e:
+                        _logger.debug("", exc_info=True)
+                        raise MlflowException.invalid_parameter_value(
+                            "Failed to validate serving input example. Please ensure that the "
+                            "input example is valid and the model's predict() method can accept "
+                            "it. Got error when calling predict() on the serving input example "
+                            f"`{serving_input}`: {e}",
```

Comment:
Some (probably many) ppl use MLflow to track experiments but don't deploy to either OSS server or Databricks. Should we downgrade this to warning, or at least letting users know how to bypass this? (remove input example and manually log signature instead if required)

Reply:
Yes I'll update the error message to include how to avoid it. But whether we want to downgrade this to warning, @BenWilson2 WDYT? If logging models to UC is required soon, then warning is not enough

Reply:
I like the idea of a warning here to not break existing workflows. Can we update the warning with instructions on post-hoc applying the correct signature and the steps that a user needs to take to validate prior to deployment?